File: lib/libraries/crypto/encryption.ex
defmodule Encryption do
  #  1. eph_pub
	#  2. sym_enc( diffie_shared_key(eph_priv, receiver_pub), sender_pub || sign(sender_priv, eph_pub) || message )

	def diffie_shared_key(priv, pub) do CryptoSign.shared_secret(pub, priv) end
	def si(key) do
		key = :crypto.hmac(:sha256, "", key)
    b = <<182, 19, 103, 154, 8, 20, 217, 236, 119, 47, 149, 215, 120, 195, 95, 197>> 
		:crypto.stream_init(:aes_ctr, key, b)
	end
	def sym_enc(key, msg) do
    {_, x} = :crypto.stream_encrypt(si(key), PackWrap.pack(msg))
		Base.encode64(x)
	end
	def sym_dec(key, e_msg) do
		{:ok, b} = Base.decode64(e_msg)
		{_, x} = :crypto.stream_decrypt(si(key), b)
		x |> PackWrap.unpack
	end
	def send_msg(to_pub, msg) do
		{eph_pub, eph_priv} = CryptoSign.new_key
		sig = %{from: Keys.pubkey,
						sig: Keys.raw_sign(eph_pub),
						msg: msg}
		%{key: eph_pub,
			msg: sym_enc(diffie_shared_key(eph_priv, to_pub), sig)}
	end
	def recieve_msg(msg) do
		sig = sym_dec(Keys.shared_secret(msg.key), msg.msg)
		if CryptoSign.verify(msg.key, sig[:sig], sig[:from]) do
			%{msg: sig[:msg], from: sig[:from]}
		else
			false
		end
	end
	def test do
		send_msg(Keys.pubkey, "test") |> recieve_msg
	end
end

File: erlang/src/encryption.erl
-module(encryption).
-export([test/0,bin_enc/2,bin_dec/2,send_msg/2,recieve_msg/1]).
-record(msg, {from = "", sig = "", msg = ""}).
-record(emsg, {key = "", msg = ""}).
si(Key) -> crypto:stream_init(rc4, crypto:hmac(sha256, "", Key)).
bin_enc(Key, Bin) ->
    {_, X} = crypto:stream_encrypt(si(Key), Bin),
    X.
bin_dec(Key, Msg) ->
    {_, Y} = crypto:stream_decrypt(si(Key), Msg),
    Y.
sym_enc(Key, Msg) -> bin_enc(Key, packer:pack(Msg)).
sym_dec(Key, Emsg) -> packer:unpack(bin_dec(Key, Emsg)).
send_msg(M, ToPub) -> 
    {EphPub, EphPriv} = sign:new_key(),
    Msg = #msg{from=keys:pubkey(), sig=keys:raw_sign(EphPub), msg=M},
    SS = sign:shared_secret(ToPub, EphPriv),
    Emsg = sym_enc(SS, Msg),
    #emsg{key=EphPub, msg=base64:encode(Emsg)}.
recieve_msg(Msg) ->
    Sig = sym_dec(keys:shared_secret(Msg#emsg.key), base64:decode(Msg#emsg.msg)),
    true = sign:verify_sig(Msg#emsg.key, Sig#msg.sig, Sig#msg.from),
    Sig.
test() ->
    Val = <<"1234">>,
    Binary = <<2,3,4>>,
    true = bin_dec("abc", bin_enc("abc", Val)) == Val,
    true = bin_dec("abc", bin_enc("abc", Binary)) == Binary,
    Record = {f, Binary},
    Sig = recieve_msg(send_msg(Record, keys:pubkey())),
    true = Sig#msg.msg == Record,
    "success".
